File: mantra_dev/backend_code/department/department.py


# import frappe
# import json

# @frappe.whitelist()
# def handle_warehouse_changes(added_warehouse, removed_warehouse, doc_name):
#     if isinstance(added_warehouse, str):
#         added_warehouse = json.loads(added_warehouse)
#     if isinstance(removed_warehouse, str):
#         removed_warehouse = json.loads(removed_warehouse)

#     if isinstance(doc_name, str):
#         doc_name = json.loads(doc_name)

#     result = {}





#     if removed_warehouse:
#         dp_records = frappe.get_all(
#             "Departmental Permission",
#             filters={"department": doc_name},
#             pluck="name"
#         )

#         for dp_name in dp_records:
#             dp_doc = frappe.get_doc("Departmental Permission", dp_name)

#             child_warehouses = dp_doc.get("warehouse")

#             updated_warehouses = [row for row in child_warehouses if row.warehouse not in removed_warehouse]

#             dp_doc.set("warehouse", updated_warehouses)

#             dp_doc.save() 

#         result["removed_warehouse"] = removed_warehouse



#     if added_warehouse:
#         result["added_warehouse"] = added_warehouse


#         dp_records1 = frappe.get_all("Departmental Permission", {"department": doc_name}, pluck="name")
      

#         for dp_name in dp_records1:
#             dp_doc = frappe.get_doc("Departmental Permission", dp_name)



           
#             for warehouse in added_warehouse:
#                 if not frappe.db.exists("Departmental Permission Warehouse",{"warehouse":warehouse,"parent":dp_doc.name,"parenttype": "Departmental Permission"}) and dp_doc.user:
#                     dp_doc.append("warehouse", {"warehouse": warehouse})

#             dp_doc.save()  

           

#     return result








# import frappe
# @frappe.whitelist()
# def handle_stock_entry_type_changes(added_stock_entry_type, removed_stock_entry_type, doc_name):
#     if isinstance(added_stock_entry_type, str):
#         added_stock_entry_type = json.loads(added_stock_entry_type)
#     if isinstance(removed_stock_entry_type, str):
#         removed_stock_entry_type = json.loads(removed_stock_entry_type)

#     if isinstance(doc_name, str):
#         doc_name = json.loads(doc_name)

#     result = {}


#     employee_user = frappe.db.get_all("Employee",{"department":doc_name},pluck='user_id')


#     if added_stock_entry_type and employee_user:
        

#         for i in added_stock_entry_type:
#             for j in employee_user:
#                 if j  and not frappe.db.exists("User Permission",{"user":j,"allow":"Stock Entry Type","for_value":i}):
#                     doc = frappe.new_doc("User Permission")
#                     doc.user = j	
#                     doc.allow = "Stock Entry Type"
#                     doc.for_value = i
#                     doc.save()



#     if removed_stock_entry_type:
#         user_perm_list = frappe.get_all(
#             "User Permission",
#             filters={
#                 "user": ["in", employee_user], 
#                 "for_value": ["in", removed_stock_entry_type],
#                 "allow":"Stock Entry Type"
#             },
#             pluck="name"
#         )

#         for i in user_perm_list:
#             frappe.delete_doc("User Permission", i)
#             frappe.db.commit()



# @frappe.whitelist()
# def material_request_type_list(department,purpose):
#     d_user = frappe.db.get_all("Departmental Permission",{"department":department},pluck="user")
#     for i in d_user:
#         doc = frappe.get_doc("Departmental Permission",i)

#         if not frappe.db.exists("Material Request Type Purpose",{"parent":i,"parenttype":"Departmental Permission","material_request_type":purpose}) and doc.user:
#             doc.append("material_request_type", {"material_request_type": purpose})
#         doc.save()
#         frappe.db.commit()



        





# @frappe.whitelist()
# def remove_material_request_type(department, purpose):
#     d_user = frappe.db.get_all("Departmental Permission", {"department": department}, pluck="user")
    
#     for i in d_user:
#         doc = frappe.get_doc("Departmental Permission", i)

#         for row in doc.get("material_request_type"):
#             if row.material_request_type == purpose:
#                 doc.remove(row)
#                 doc.save()
                
#                 frappe.db.commit()
    

File: mantra_dev/public/js/department.js



// frappe.ui.form.on("Department", {
//     onload: function (frm) {
//         frm.previous_warehouse = (frm.doc.custom_departmental_warehouse || []).map(row => row.warehouse);
//         frm.previous_stock_entry_type = (frm.doc.custom_departmental_stock_entry_type || []).map(row => row.stock_entry_type);

//     },

//     before_save: function (frm) {
//         let current_items = (frm.doc.custom_departmental_warehouse || []).map(row => row.warehouse);

//         let removed_warehouse = frm.previous_warehouse.filter(warehouse => !current_items.includes(warehouse));
//         let added_warehouse = current_items.filter(warehouse => !frm.previous_warehouse.includes(warehouse));




//         let current_stock_entry_type = (frm.doc.custom_departmental_stock_entry_type || []).map(row => row.stock_entry_type);

//         let removed_stock_entry_type = frm.previous_stock_entry_type.filter(stock_entry_type => !current_stock_entry_type.includes(stock_entry_type));
//         let added_stock_entry_type = current_stock_entry_type.filter(stock_entry_type => !frm.previous_stock_entry_type.includes(stock_entry_type));
//         console.log(added_stock_entry_type)

       

//         if (removed_warehouse.length > 0 || added_warehouse.length > 0) {
//             frappe.call({
//                 method: "mantra_dev.backend_code.department.department.handle_warehouse_changes",
//                 args: {
//                     added_warehouse: JSON.stringify(added_warehouse),  // Convert to JSON string
//                     removed_warehouse: JSON.stringify(removed_warehouse),
//                     doc_name:JSON.stringify(frm.doc.name)
//                 },
//                 callback: function(response) {
//                     if (response.message) {

//                         if (response.message.removed_warehouse) {
//                         }
//                         if (response.message.added_warehouse) {
//                         }
//                     } else {
//                     }
//                 }
//             });
//         }

//         if (removed_stock_entry_type.length > 0 || added_stock_entry_type.length > 0) {
//             frappe.call({
//                 method: "mantra_dev.backend_code.department.department.handle_stock_entry_type_changes",
//                 args: {
//                     added_stock_entry_type: JSON.stringify(added_stock_entry_type),  // Convert to JSON string
                    
//                     removed_stock_entry_type: JSON.stringify(removed_stock_entry_type),
//                     doc_name:JSON.stringify(frm.doc.name)
//                 },
                
//                 callback: function(response) {
//                     if (response.message) {

//                         if (response.message.removed_warehouse) {
//                         }
//                         if (response.message.added_warehouse) {
//                         }
//                     } else {
//                     }
//                 }
//             });
//         }


//         frm.previous_warehouse = [...current_items];
//         frm.previous_stock_entry_type = [...current_stock_entry_type];
//     },

    

// });





// frappe.ui.form.on('Material Request Type Purpose', {
//     material_request_type: function(frm, cdt, cdn) {
//         let row = locals[cdt][cdn];
//         console.log(row.parent)

//         // Ensure department exists before making a request
//         if (row.material_request_type) {
//             frappe.call({
//                 method: "mantra_dev.backend_code.department.department.material_request_type_list",
//                 args: {
//                     department: row.parent,
//                     purpose:row.material_request_type
//                 },
//                 // console.log(row.parent);
//                 callback: function(response) {
//                     if (response.message) {
//                         console.log("Fetched Departmental Permission Records:", response.message);
//                         frappe.msgprint(__("Departmental Permission records fetched. Check console."));
//                     }
//                 }
//             });
//         }
//     },
//     before_custom_material_request_type_remove: function(frm, cdt, cdn) {
//         let row = locals[cdt][cdn];
//         console.log(row.parent)

//         frappe.call({
//             method: "mantra_dev.backend_code.department.department.remove_material_request_type",
//             args: {
//                 department: row.parent,
//                 purpose: row.material_request_type
//             },
//             callback: function(response) {
//                 if (response.message) {
//                     console.log("Removed Departmental Permission Record:", response.message);
//                     frappe.msgprint(__("Departmental Permission record removed."));
//                 }
//             }
//         });
//     }
    
// });


